File: java/terasort/1_base/src/main/java/terasort/filesFiltered2/Sort.java
/*
 *  Copyright 2002-2016 Barcelona Supercomputing Center (www.bsc.es)
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package terasort.filesFiltered2;

import java.io.File;
import java.util.LinkedList;

public class Sort {

    private static int NUM_FRAGMENTS;
    private static int NUM_RANGES;
    private static long minKey;
    private static long maxKey;
    private static String DATA_FOLDER;
    private static String OUTPUT_FOLDER;

    private static String[] filePaths;
    //private static Fragment[] result;

    public static void main(String args[]) {
        // Get parameters
        if (args.length != 3) {
            System.out.println("[ERROR] Usage: Sort <DATA_FOLDER> <OUTPUT_FOLDER> <NUM_RANGES>");
            System.exit(-1);
        }
        DATA_FOLDER = args[0];
        System.out.println("[LOG] DATA_FOLDER   = " + DATA_FOLDER);
        OUTPUT_FOLDER = args[1];
        System.out.println("[LOG] OUTPUT_FOLDER = " + OUTPUT_FOLDER);
        NUM_RANGES = Integer.parseInt(args[2]);
        System.out.println("[LOG] NUM_RANGES = " + NUM_RANGES);

        // Run sort by key app
        long startTime = System.currentTimeMillis();
        Sort sbk = new Sort();
        sbk.run();

        System.out.println("[LOG] Main program finished.");

        // Uncomment the following lines to see the result keys
        //for (int i=0; i < NUM_RANGES; ++i){
        //	System.out.println("Range " + i + " result:");
        //	System.out.println(result[i]);
        //}
        long endTime = System.currentTimeMillis();

        System.out.println("[TIMER] Elapsed time: " + (endTime - startTime) + " ms");
    }

    /*
	 * This run reads only the range that requires from each file.
	 * Avoids to transfer the entire input file.
     */
    private void run() {
        // Initialize file Names
        System.out.println("[LOG] Initialising filenames for each matrix");
        initializeVariables();

        // Compute result
        System.out.println("[LOG] Reading fragments.");
        System.out.println("[LOG] Filtering fragments.");

        Fragment[][] fragments = new Fragment[NUM_FRAGMENTS][NUM_RANGES];
        Range[][] ranges = new Range[NUM_FRAGMENTS][NUM_RANGES];
        long step = ((maxKey - minKey) / NUM_RANGES);

        for (int i = 0; i < NUM_FRAGMENTS; ++i) {
            long start = 0;
            long end = step;
            for (int j = 0; j < NUM_RANGES; ++j) {
                fragments[i][j] = SortImpl.getFilteredFragment(filePaths[i], start, end);
                ranges[i][j] = new Range(start, end);
                start += step;
                end += step;
            }
        }

        System.out.println("[LOG] Merging filtered results...");
        System.out.println("[LOG] Sorting fragment...");
        System.out.println("[LOG] Saving sorted fragments...");

        int nf = NUM_FRAGMENTS;
        int nr = NUM_RANGES;
        Range[][] aux;
        Fragment[][] auxFragments;
        while (nf > 1) {
            nf = nf / 2;
            nr = nr * 2;
            aux = new Range[nf][nr];
            auxFragments = new Fragment[nf][nr];
            for (int i = 0; i < nf; i++) {
                for (int j = 0; j < nr; j++) {
                    auxFragments[i][j] = new Fragment();
                }
            }
            int posX = 0;
            for (int i = 0; i < nf * 2; i += 2) {
                int posY = 0;
                for (int j = 0; j < nr / 2; j++) {
                    long s = ranges[i][j].getStart();
                    long e = ranges[i + 1][j].getEnd();
                    Range r = new Range(s, e);
                    SortImpl.mixFragments(fragments[i][j], fragments[i + 1][j], r, auxFragments[posX][posY], auxFragments[posX][posY + 1]);
                    long h = s + ((e - s) / 2);
                    aux[posX][posY] = new Range(s, h);
                    aux[posX][posY + 1] = new Range(h, e);
                    posY += 2;
                }
                posX++;
            }
            ranges = aux;
            fragments = auxFragments;
        }
        Integer[] sortedCount = new Integer[fragments[0].length];

        // sort all fragments
        for (int i = 0; i < fragments[0].length; i++) {
            Fragment sortedFragment = SortImpl.sortPartition(fragments[0][i]);
            sortedCount[i] = SortImpl.saveFragment(sortedFragment, OUTPUT_FOLDER);
        }

        // Merge-Reduce counters
        LinkedList<Integer> q = new LinkedList<Integer>();
        for (int j = 0; j < fragments[0].length; ++j) {
            q.add(j);
        }
        int x = 0;
        while (!q.isEmpty()) {
            x = q.poll();
            int y;
            if (!q.isEmpty()) {
                y = q.poll();
                sortedCount[x] = SortImpl.reduceCount(sortedCount[x], sortedCount[y]);
                q.add(x);
            }
        }
        int count = sortedCount[x];

        System.out.println("[TERASORT] Total Sorted Elements = " + count);

        System.out.println("[TERASORT] FINISHED");
    }

    private void initializeVariables() {
        NUM_FRAGMENTS = new File(DATA_FOLDER).listFiles().length;
        filePaths = new String[NUM_FRAGMENTS];
        int i = 0;
        for (File f : new File(DATA_FOLDER).listFiles()) {
            filePaths[i] = f.getAbsolutePath();
            System.out.println("File: " + filePaths[i]);
            i = i + 1;
        }

        //result = new Fragment[NUM_FRAGMENTS];
        byte[] mi = {0, 0, 0, 0, 0, 0, 0, 0};
        minKey = Utils.byte2long(mi);
        byte[] ma = {0, -1, -1, -1, -1, -1, -1, -1}; // 0xff = -1
        maxKey = Utils.byte2long(ma);
        System.out.println("Min key: " + minKey);
        System.out.println("Max key: " + maxKey);
    }

}

File: python/examples/sort_by_key/src/sort.py
#!/usr/bin/python
#
#  Copyright 2002-2019 Barcelona Supercomputing Center (www.bsc.es)
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#

# -*- coding: utf-8 -*-

from pycompss.api.task import task


def chunks(l, n, balanced=False):
    if not balanced or not len(l) % n:
        for i in range(0, len(l), n):
            yield l[i:i + n]
    else:
        rest = len(l) % n
        start = 0
        while rest:
            yield l[start: start + n + 1]
            rest -= 1
            start += n + 1
        for i in range(start, len(l), n):
            yield l[i:i + n]


def partition_by(self, num_partitions, hash):
    if not hash:
        print("Num lines: " + str(len(self)) + ", Num patitions: " + str(num_partitions))
        return chunks(self, int(len(self) / num_partitions), True)
    else:
        partitions = [[] for n in range(num_partitions)]
        for s in self:
            partitions[hashed_partitioner(s, num_partitions)].append(s)
        return partitions


def hashed_partitioner(k, num_partitions):
    return hash(keyfunc(k)) % num_partitions


def keyfunc(x):
    return x


@task(returns=list)
def sort_partition(iterator, ascending=True):
    """Sorts self, which is assumed to consists of (key, value) pairs"""
    return sorted(iterator,
                  key=lambda k_v: keyfunc(k_v[0]),
                  reverse=not ascending)


def sort_by_key(self, num_partitions=None):
    return list(map(sort_partition, partition_by(self, num_partitions, False)))


def main():
    import sys
    from pycompss.api.api import compss_wait_on

    if len(sys.argv) != 2:
        print("Usage: sort <INPUT_FILE>")
    input_path = sys.argv[1]

    text = open(input_path)
    lines = []

    for line in text:
        lines += [(x, 1) for x in line.strip().split(" ")]
    text.close()

    default_parallelism = 4
    reducer = int(max(default_parallelism / 2, 2))

    result = sort_by_key(lines, num_partitions=reducer)

    result = compss_wait_on(result)

    print([x[0] for x in result])


if __name__ == "__main__":
    main()
